[Enhancement] Two-handed sword spin attacks

Eblo opened the pull request:
This establishes parity with the magic spin attack for the Fierce Deity sword and the Great Fairy's Sword. Do note that I had to do some refactoring in `z_player.c` to make the GFS work, as there are various points where only the B button was checked for input, while the GFS is a C-button item.

https://github.com/HarbourMasters/2ship2harkinian/assets/7004497/cc63727c-38cd-4009-9a71-52045088b3d2



<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/2041890059.zip)
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/2041903907.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/2041944009.zip)
<!--- section:artifacts:end -->

smashbro596 commented:
this is godlike! this was one of my big asks! have fd link be able to use the spin attack!

Eblo committed:
Enhancement to enable magic for 2-handed swords

Create a new magic charge size for the two-handed swords
Add utility method for determining if the held item's button is being pressed, not just B

garrettjoecox commented on mm/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -8185,14 +8187,33 @@ s32 func_8083A4A4(Player* this, f32* arg1, s16* arg2, f32 arg3) {
     return false;
 }
 
+// #region 2S2H [Port] [Dpad] Convenience method that checks any held item button, not just B
+bool isPressingHeldItemButton(Player* this) {
+    if (this->heldItemButton < 0) {
+        return false;
+    }
+
+    uint16_t buttonToCheck;
+    if (IS_HELD_DPAD(this->heldItemButton)) {
+        buttonToCheck = sDpadItemButtons[HELD_ITEM_TO_DPAD(this->heldItemButton)];
+    } else {
+        buttonToCheck = sPlayerItemButtons[this->heldItemButton];
+    }
+
+    return CHECK_BTN_ALL(sPlayerControlInput->cur.button, buttonToCheck);
+}
+// #endregion
```
It looks like this function is only ever called from your enhancement file in C++, can it just be moved there?

Eblo replied:
Sure, I'm not against that. In my head I thought this might be something others may want to use so I left it in z_player.c, but since there is nobody calling for it now, that's silly. I'll move it.

garrettjoecox commented on mm/src/overlays/actors/ovl_En_M_Thunder/z_en_m_thunder.c:
```diff
@@ -547,7 +547,15 @@ void EnMThunder_Draw(Actor* thisx, PlayState* play2) {
         Matrix_RotateXS(0x4000, MTXMODE_APPLY);
-    } else {
+    }
+    // #region 2S2H [Enhancements] Two Handed Sword Magic Charge Effect
```
Seems harmless enough, it's rather difficult doing stuff like this from hooks. Let's wrap it in a cvar at least though?

Eblo replied:
It actually wasn't too troublesome to hookify this. There exists a visual bug in the original game where the player could get the GFS to do a charged magic spin attack, but it would display the Kokiri Sword's visual effect. The way I did this enhancement previously had fixed that effect regardless of whether the enhancement was enabled or disabled. With the new hook, the effect is only fixed with the enhancement on, so purists can still see the original bug if the enhancement is disabled.

garrettjoecox commented on mm/2s2h/Enhancements/GameInteractor/GameInteractor.h:
```diff
@@ -55,6 +55,8 @@ typedef enum {
     GI_VB_FLIP_HOP_VARIABLE,
     GI_VB_DISABLE_LETTERBOX,
     GI_VB_CLOCK_TOWER_OPENING_CONSIDER_THIS_FIRST_CYCLE,
+    GI_VB_HELD_ITEM_BUTTON_PRESS,
+    GI_VB_MAGIC_SPIN_ATTACK_FORM_CHECK,
```
Both of these could be a bit better, should read as a question (Should held item button press?) maybe "Should held item consider button pressed?" Bit of a mouthful but it's clearer

For the other you can just swap the last two words "Should magic spin attack check form?"

Eblo replied:
I renamed `GI_VB_HELD_ITEM_BUTTON_PRESS` to `VB_CHECK_HELD_ITEM_BUTTON_PRESS`. Vanilla behavior is to *only* check B, regardless of what button the item is actually assigned to. Incidentally, this means vanilla behavior is that the GFS can do a charged magic spin attack if the player quickly holds B after swinging the GFS. What this hook does is instead check the currently held item's assigned button, which may be B or some other button. From what I tested, that also means that, not only does this enhancement render the GFS magic charge glitch pointless, but it is also not possible when enabled. The glitch can still be performed when the enhancement disabled.

garrettjoecox reviewed (approved):
a few comments but mostly looks good, needs some updating

Eblo committed:
Clean up two-handed sword spin attacks

Add hook for thunder matrix transformation, move that logic into TwoHandedSwordSpinAttack
Move held item button press function into TwoHandedSwordSpinAttack.cpp
Add enhancement to the searchable menu
Clean up logic for magic spin attack form check
Use variadic args for hooks

Eblo committed:
Remove redundant includes from TwoHandedSwordSpinAttack.cpp

Eblo committed:
Use extern on TwoHandedSwordSpinAttack structs

Eblo commented:
I addressed the feedback. For players, the main change of note is that the new two-handed sword magic charge visual effect will only display when this enhancement is enabled.
